File: leetcode/1942-The-Number-of-the-Smallest-Unoccupied-Chair/java/Solution.java
public class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        int n = times.length;
        int[][] orderedTimes = new int[n][2];
        for (int i = 0; i < n; i++) {
            orderedTimes[i] = times[i];
        }
        Arrays.sort(orderedTimes, (a, b) -> Integer.compare(a[0], b[0]));

        PriorityQueue<Integer> emptySeats = new PriorityQueue<>();
        for (int i = 0; i < n; i++) {
            emptySeats.offer(i);
        }

        PriorityQueue<int[]> occupiedSeats = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        for (int i = 0; i < n; i++) {
            int ar = orderedTimes[i][0];
            int lv = orderedTimes[i][1];
            while (!occupiedSeats.isEmpty() && occupiedSeats.peek()[0] <= ar) {
                emptySeats.offer(occupiedSeats.peek()[1]);
            }
            int seat = emptySeats.poll();
            occupiedSeats.offer(new int[]{lv, seat});
        }
        return -1;
    }
}

File: leetcode/1942-The-Number-of-the-Smallest-Unoccupied-Chair/kotlin/Solution.kt
import java.util.PriorityQueue

class Solution {
    fun smallestChair(times: Array<IntArray>, targetFriend: Int): Int {
        val n = times.size
        val orderedTimes = times.copyOf().sortedWith(Comparator.comparingInt { it[0] })

        val emptySeats = PriorityQueue<Int> { a, b -> a - b }
        for (i in 0 until n) {
            emptySeats.add(i)
        }
        val takenSeats = PriorityQueue<IntArray>(Comparator.comparingInt { it[0] })

        for (i in 0 until n) {
            val (ar, lv) = times[i]
            while (takenSeats.isNotEmpty() && takenSeats.peek()[0] <= ar) {
                emptySeats.add(takenSeats.poll()[1])
            }
            val seat = emptySeats.poll()
            if (i == targetFriend) return seat
            takenSeats.offer(intArrayOf(lv, seat))
        }
        return -1
    }
}